singer_sdk/testing/factory.py

```diff
@@ -0,0 +1,240 @@
+"""Test Class Factory."""
+
+from __future__ import annotations
+
+from typing import Any, cast
+
+import pytest
+
+from singer_sdk import Tap, Target
+
+from .config import SuiteConfig
+from .runners import TapTestRunner, TargetTestRunner
+from .suites import (
+    tap_stream_attribute_tests,
+    tap_stream_tests,
+    tap_tests,
+    target_tests,
+)
+
+
+def get_test_class(
+    test_runner: TapTestRunner | TargetTestRunner,
+    test_suites: list,
+    suite_config: SuiteConfig | None,
+) -> object:
+    """Construct a valid pytest test class from given suites.
+
+    Args:
+        test_runner: A Tap or Target test runner instance.
+        test_suites: A list of Test Suits to apply.
+        suite_config: SuiteConfig instance to pass to tests.
+
+    Returns:
+        A test class usable by pytest.
+    """
+    suite_config = suite_config or SuiteConfig()
+
+    class BaseTestClass:
+        """Base test class."""
+
+        params: dict = {}
+        param_ids: dict = {}
+
+        @pytest.fixture
+        def config(self) -> SuiteConfig:
+            return suite_config or SuiteConfig()
+
+        @pytest.fixture
+        def resource(self) -> Any:  # noqa: ANN401
+            yield
+
+        @pytest.fixture(scope="class")
+        def runner(self) -> TapTestRunner | TargetTestRunner:
+            # Populate runner class with cached records for use in tests
+            test_runner.sync_all()
+            return test_runner
+
+    for suite in test_suites:
+
+        # make sure given runner is of type TapTestRunner
+        expected_runner_class = (  # type: ignore[valid-type]
+            TapTestRunner
+            if suite.kind in {"tap", "tap_stream", "tap_stream_attribute"}
+            else TargetTestRunner
+        )
+        assert isinstance(test_runner, expected_runner_class), (
+            f"Test suite of kind {suite.kind} passed, "
+            f"but test runner if of type {type(test_runner)}."
+        )
+        test_runner = cast(
+            expected_runner_class, test_runner  # type: ignore[valid-type]
+        )
+
+        if suite.kind in {"tap", "target"}:
+            for TestClass in suite.tests:
+                test = TestClass()
+                test_name = f"test_{suite.kind}_{test.name}"
+                setattr(BaseTestClass, f"test_{suite.kind}_{test.name}", test.run)
+
+        if suite.kind in {"tap_stream", "tap_stream_attribute"}:
+
+            streams = list(test_runner.tap.streams.values())
+            if suite_config.max_records_limit:
+                for stream in streams:
+                    stream._MAX_RECORDS_LIMIT = suite_config.max_records_limit
```

Comment:
@kgpayne I'm seeing the tests running for as long as before (~7.5 min) so maybe this isn't really working to limit the number of records.

Reply:
@edgarrmondragon you are right - this is only applying it to the `tap` instance used to construct tests - not the instance created by the runner at run time. However I don't believe the `_MAX_RECORDS_LIMIT` can be used in this way either 😔 Will push failing tests so you can see what I mean.

Reply:
@kgpayne I see. You're right, the limit was only used in the connection test. So:

Do you think it's worth the level of effort to implement a limit? (perhaps by catching `MaxRecordsLimitException` in `TapTestRunner._execute_sync`)

Reply:
@edgarrmondragon yeh, I thought about trying to catch `MaxRecordsLimitException` but i) I think it would be lots of copied logic (as you say) and ii) I still think the issue is upstream of the check; we want `Stream._sync_records()` to exit gracefully when the limit is reached (emitting a last state message etc.) rather than raising. This either means changing `_check_max_record_limit()` to return a boolean that exits the loop through `get_records()` _or_ we want `get_records()` to only return `_MAX_LIMIT_RECORDS` records so the error is never raised.

We took the latter approach in the `SQLStream` class [here](https://github.com/meltano/sdk/blob/main/singer_sdk/streams/sql.py#L198) - opting to "push down" the limit onto the underlying SQL engine, rather than generating a full result-set on the server but returning early in the client/tap.

So I think a solution would be to update `RESTStream.get_records()` in the same way to return gracefully if the limit is hit, especially if it means fetching fewer pages etc. 🙂 WDYT?

Reply:
@aaronsteers FYI as this has come up after the update I gave in our 1:1 👆 I still think @edgarrmondragon is right in saying we should avoid doing full syncs during automated testing, but I also don't want to expand scope too much in cases where other parts of the code may need to change to facilitate that as a feature 😅

Reply:
> So I think a solution would be to update `RESTStream.get_records()` in the same way to return gracefully if the limit is hit, especially if it means fetching fewer pages etc. 🙂 WDYT?

@kgpayne It's probably not worth the effort of messing with the generator in that method. 

Mutating  the `_MAX_LIMIT_RECORDS` class attribute feels clunky anyway and I'd prefer something like `Tap.sync(max_records=5)`, but I'm ok with making that a follow-up and remove the option from `SuiteConfig` in this PR.

Reply:
@edgarrmondragon OK, great. I have created a follow-up [issue](https://github.com/meltano/sdk/issues/1349), and commented [here](https://github.com/meltano/sdk/issues/1333#issuecomment-1403751140) in a related issue ✅